chain/db/name.go

```diff
@@ -7,96 +7,153 @@ import (
 	"github.com/nknorg/nkn/common/serialization"
 )
 
-func generateRegistrantKey(registrant []byte) []byte {
-	registrantKey := bytes.NewBuffer(nil)
-	registrantKey.WriteByte(byte(NS_Registrant))
-	serialization.WriteVarBytes(registrantKey, registrant)
-	return registrantKey.Bytes()
+func getRegistrantId(registrant []byte) string {
+	return string(registrant)
 }
 
-func generateNameKey(name string) []byte {
-	nameKey := bytes.NewBuffer(nil)
-	nameKey.WriteByte(byte(NS_Name))
-	serialization.WriteVarString(nameKey, strings.ToLower(name))
-	return nameKey.Bytes()
+func getNameId(name string) string {
+	return strings.ToLower(name)
 }
 
-func (cs *ChainStore) SaveName(registrant []byte, name string) error {
-	registrantKey := generateRegistrantKey(registrant)
-	nameKey := generateNameKey(name)
+func (sdb *StateDB) updateName(registrant []byte, name string) error {
+	registrantId := getRegistrantId(registrant)
+	nameId := getNameId(name)
 
-	// PUT VALUE
 	w := bytes.NewBuffer(nil)
 	serialization.WriteVarString(w, name)
-	err := cs.st.BatchPut(registrantKey, w.Bytes())
+	err := sdb.trie.TryUpdate(append(NamePrefix, nameId...), w.Bytes())
 	if err != nil {
 		return err
 	}
 
 	w = bytes.NewBuffer(nil)
 	serialization.WriteVarBytes(w, registrant)
-	err = cs.st.BatchPut(nameKey, w.Bytes())
-	if err != nil {
-		return err
-	}
-
-	return nil
+	return sdb.trie.TryUpdate(append(NameRegistrantPrefix, registrantId...), w.Bytes())
 }
 
-func (cs *ChainStore) DeleteName(registrant []byte) error {
-	name, err := cs.GetName(registrant)
-	if err != nil {
-		return err
-	}
+func (sdb *StateDB) setName(registrant []byte, name string) {
+	registrantId := getRegistrantId(registrant)
+	nameId := getNameId(name)
 
-	registrantKey := generateRegistrantKey(registrant)
-	nameKey := generateNameKey(*name)
+	sdb.names[registrantId] = name
+	sdb.nameRegistrants[nameId] = registrant
+}
+
+func (cs *ChainStore) SetName(registrant []byte, name string) {
+	cs.States.setName(registrant, name)
+}
 
-	// DELETE VALUE
-	err = cs.st.BatchDelete(registrantKey)
+func (sdb *StateDB) deleteName(registrantId string, nameId string) error {
+	err := sdb.trie.TryDelete(append(NameRegistrantPrefix, registrantId...))
 	if err != nil {
 		return err
 	}
-	err = cs.st.BatchDelete(nameKey)
+
+	err = sdb.trie.TryDelete(append(NamePrefix, nameId...))
```

Comment:
What if the previous TryDelete succeeded but this one failed?

Reply:
we'll get incorrect state that won't be saved I guess. not sure what can be done about that.

Reply:
I think it's better to provide some batch operation API in trie later